plugins/inputs/t128_tank/reader.go

```diff
@@ -125,42 +125,47 @@ func (r *Reader) withTankReadCommandContext(tankReadCmdCtx CommandContext) *Read
 	return r
 }
 
+func (r *Reader) saveIndex(indexValue uint64) {
+	r.lastSavedIndex = indexValue
+	r.setIndex(r.indexPath, index{value: indexValue})
+}
+
 func (r *Reader) Run(mainCtx context.Context) {
 	readCtx, readCtxCancel := context.WithCancel(mainCtx)
 	lastIndex, err := r.getIndex(r.indexPath, r.defaultIndex)
-	r.lastSavedIndex = lastIndex.value
 	if err != nil {
 		r.log.Errorf("Error in get index %v", err)
 		readCtxCancel()
 		return
 	}
 
-	nextSaveCheck := time.After(2 * time.Second)
-	defer func() {
-		readCtxCancel()
-		r.setIndex(r.indexPath, index{value: r.lastSavedIndex})
-	}()
-
+	nextSaveCheck := time.NewTicker(2 * time.Second)
+	defer readCtxCancel()
 	go r.read(readCtx, lastIndex.next())
 
 	for {
+
 		select {
 		case <-mainCtx.Done():
 			r.log.Errorf("%s reader done", r.topic)
+			r.saveIndex(r.lastSavedIndex)
 			return
-		case <-nextSaveCheck:
+		case <-nextSaveCheck.C:
 			if lastIndex.value > r.lastSavedIndex {
-				r.setIndex(r.indexPath, lastIndex)
-				r.lastSavedIndex = lastIndex.value
+				r.saveIndex(lastIndex.value)
+			}
+			if lastIndex.value%1000 == 0 {
```

Comment:
This needs to be done when the message is sent, not on the tick timer. At this point, the index would have already been saved because of the above if statement. Something will need to indicate to this thread that an index was processed (probably the reader). Alternatively, we can just rely on the exit functionality. But I'd prefer if we're proactive and save every N (1000) messages.